adapters/zio-http/src/main/scala/caliban/ZHTTPAdapter.scala

```diff
@@ -0,0 +1,240 @@
+  }
+
+  private def generateGraphQLResponse[R, E](
+    payload: GraphQLRequest,
+    id: String,
+    interpreter: GraphQLInterpreter[R, E],
+    skipValidation: Boolean,
+    enableIntrospection: Boolean,
+    queryExecution: QueryExecution,
+    subscriptions: Subscriptions
+  ): ZStream[R, E, Text] =
+    ZStream
+      .fromEffect(
+        interpreter
+          .executeRequest(payload, skipValidation, enableIntrospection, queryExecution)
+      )
+      .flatMap(res =>
+        res.data match {
+          case ObjectValue((fieldName, StreamValue(stream)) :: Nil) =>
+            trackSubscription(id, subscriptions).flatMap { p =>
+              stream.map(toResponse(id, fieldName, _, res.errors)).interruptWhen(p)
+            }
+          case other                                                =>
+            ZStream.succeed(toResponse(id, GraphQLResponse(other, res.errors)))
+        }
+      )
+      .catchAll { _ =>
+        connectionError
+      }
+
+  private def executeToJson[R, E](
+    interpreter: GraphQLInterpreter[R, E],
+    request: GraphQLRequest,
+    skipValidation: Boolean,
+    enableIntrospection: Boolean,
+    queryExecution: QueryExecution
+  ): URIO[R, Json] =
+    interpreter
+      .executeRequest(
+        request,
+        skipValidation = skipValidation,
+        enableIntrospection = enableIntrospection,
+        queryExecution
+      )
+      .foldCause(cause => GraphQLResponse(NullValue, cause.defects).asJson, _.asJson)
+
+  implicit class HttpErrorOps[R, E <: Throwable, A](private val zio: ZIO[R, io.circe.Error, A]) extends AnyVal {
+    def handleHTTPError: ZIO[R, HttpError, A] = zio.mapError({
+      case DecodingFailure(error, _)  =>
+        HttpError.BadRequest.apply(s"Invalid json: $error")
+      case ParsingFailure(message, _) =>
+        HttpError.BadRequest.apply(message)
+      case t: Throwable               => HttpError.InternalServerError.apply("Internal Server Error", Some(t.getCause()))
+    })
+  }
+
+  private val protocol = SocketProtocol.subProtocol("graphql-ws")
+
+  private val keepAlive = (keepAlive: Option[Duration]) =>
+    keepAlive match {
+      case None           => ZStream.empty
+      case Some(duration) =>
+        ZStream
+          .succeed(Text("""{"type":"ka"}"""))
+          .repeat(Schedule.spaced(duration))
+    }
+
+  private val connectionError                            = ZStream.succeed(Text("""{"type":"connection_error"}"""))
+  private val connectionAck                              = ZStream.succeed(Text("""{"type":"connection_ack"}"""))
+  private val close                                      = ZStream.succeed(WebSocketFrame.close(1000))
+  private def toStreamError[E](id: Option[String], e: E) = ZStream.succeed(
+    Text(
+      Json
+        .obj(
+          "id"      -> Json.fromString(id.getOrElse("")),
+          "type"    -> Json.fromString("error"),
+          "message" -> Json.fromString(e.toString())
```

Comment:
Is this how we are doing this elsewhere?

Reply:
Pretty much like this :) https://github.com/ghostdogpr/caliban/blob/master/adapters/http4s/src/main/scala/caliban/Http4sAdapter.scala#L242